src/linear_algebra/projection.lean

```diff
@@ -330,3 +332,87 @@ def is_compl_equiv_proj :
   (p.is_compl_equiv_proj.symm f : submodule R E) = (f : E →ₗ[R] p).ker := rfl
 
 end submodule
+
+namespace linear_map
+
+open submodule
+
+/--
+A linear endomorphism of a module `E` is a projection onto a submodule `p` if it sends every element
+of `E` to `p` and fixes every element of `p`.
+-/
+structure is_proj (f : E →ₗ[R] E) : Prop :=
```

Comment:
```suggestion
structure is_proj {F : Type*} [fun_like F E (λ _, E)] (f : F) : Prop :=
```
I think it would be nice to generalize it that way. The drawback is that it requires `R` to be an explicit argument to `is_proj`, but then it would also make sense for continuous linear maps, matrices, and so on.

Reply:
What would you do with `p` then? Right now it's a `submodule R E`, should it be replaced by just a `set E` or something else?

Reply:
Actually forget what I said about needing `R` to be explicit -- it can be inferred from `p`.

Reply:
But if `p` is still a `submodule` I don't really see the point of the generalization. For matrices or continuous linear maps can't we just use `is_proj p f.to_linear_map`?

Reply:
You can just leave `p` as a submodule. `fun_like` is a bit of overkill, but `linear_map_class` is not defined yet and it wouldn't buy us anything over `fun_like`.

Avoiding `to_linear_map` is exactly the point here -- it gets old very fast.

Reply:
Sure, then